In [1]:
import numpy as np
import pandas as pd 
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

In [2]:
data = pd.read_csv('2.02. Binary predictors.csv')

In [3]:
data.head()

SAT Admitted  Gender
0  1363       No    Male
1  1792      Yes  Female
2  1954      Yes  Female
3  1653       No    Male
4  1593       No    Male

In [4]:
data['Admitted'] = data['Admitted'].map({'Yes':1,'No':0})
data['Gender'] = data['Gender'].map({'Female':1,'Male':0})
data.head()

SAT  Admitted  Gender
0  1363         0       0
1  1792         1       1
2  1954         1       1
3  1653         0       0
4  1593         0       0

In [5]:
y = data['Admitted']
x1 = data[['SAT','Gender']]

In [6]:
x = sm.add_constant(x1)
reg_log = sm.Logit(y,x)
result_log = reg_log.fit()
result_log.summary()

Optimization terminated successfully.
         Current function value: 0.120117
         Iterations 10


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:               Admitted   No. Observations:                  168
Model:                          Logit   Df Residuals:                      165
Method:                           MLE   Df Model:                            2
Date:                Sat, 09 Mar 2019   Pseudo R-squ.:                  0.8249
Time:                        14:49:36   Log-Likelihood:                -20.180
converged:                       True   LL-Null:                       -115.26
                                        LLR p-value:                 5.118e-42
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const        -68.3489     16.454     -4.154      0.000    -100.598     -36.100
SAT            0.0406      0.010      4.129      0.000       0.021       0.060
Gender         1.9449      0.846      2.299      0.022       0.287       3.603
==============================================================================

Possibly complete quasi-separation: A fraction 0.27 of observations can be
perfectly predicted. This might indicate that there is complete
quasi-separation. In this case some parameters will not be identified.
"""

## Exponential function for result

In [7]:
np.exp(1.9449)

6.992932526814459

## This means there is 7 times more chances for female candidate than men for the admission having same SAT Score

## Lets get the Admitted data in array to see how our model predict 

In [8]:
np.array(data['Admitted'])

array([0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1,
       0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0,
       1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0,
       0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0], dtype=int64)

## Lets use the pred function of statsmodal

In [9]:
result_log.pred_table()

array([[69.,  5.],
       [ 4., 90.]])

## The above table is called Confusion matrix and it shows how well our model predict

In [10]:
cf = pd.DataFrame(result_log.pred_table())
cf.columns = ['Predicted 0','Predicted 1']
cf = cf.rename(index={0:'Actual 0',1:'Actual 1'})
cf

Predicted 0  Predicted 1
Actual 0         69.0          5.0
Actual 1          4.0         90.0

## 69 Corresponding to actual value is 0 and the predicted one is also 0
## 90 Corresponding to actual value is 1 and the predicted one is also 1

## 5 Corresponding to actual value is 0 but our modal predict 1
## 4 Corresponding to actual value is 1 but our modal predict 0

## Lets calculate the accuracy of our modal

In [11]:
cm = np.array(cf)
acccuracy = ((cm[0,0]+cm[1,1])/cm.sum())
Per = acccuracy*100
Per

94.64285714285714

## Our modal accuracy is 94.6 %

## Lets test the test data 

In [12]:
test = pd.read_csv('2.03. Test dataset.csv')

In [13]:
test.head()

SAT Admitted  Gender
0  1323       No    Male
1  1725      Yes  Female
2  1762      Yes  Female
3  1777      Yes    Male
4  1665       No    Male

In [14]:
test['Admitted'] = test['Admitted'].map({'Yes':1,'No':0})
test['Gender'] = test['Gender'].map({'Female':1,'Male':0})
test.head()

SAT  Admitted  Gender
0  1323         0       0
1  1725         1       1
2  1762         1       1
3  1777         1       0
4  1665         0       0

In [15]:
test_actual = test['Admitted']
test_data = test.drop(['Admitted'],axis=1)
test_data = sm.add_constant(test_data)
test_data.head()

const   SAT  Gender
0    1.0  1323       0
1    1.0  1725       1
2    1.0  1762       1
3    1.0  1777       0
4    1.0  1665       0

In [22]:
def confusion_matrix(data,actual_data,model):
    
    pred_val = model.predict(data)
    bins = np.array([0,0.5,1])
    #to create confusion matrix
    cm = np.histogram2d(actual_data,pred_val,bins = bins)[0]
    accuracy = (cm[0,0]+cm[1,1])/cm.sum()
    return cm, accuracy

In [23]:
test_actual.head()

0    0
1    1
2    1
3    1
4    0
Name: Admitted, dtype: int64

In [24]:
cm = confusion_matrix(test_data,test_actual,result_log)
cm

(array([[ 5.,  1.],
        [ 1., 12.]]), 0.8947368421052632)

## The modal is 89.47 %accurate with 90 - 10 split data

## The Training accuracy is always better than test accuracy because of the overfitting factor

In [26]:
cf = pd.DataFrame(cm[0])
cf.columns = ['Predicted 0','Predicted 1']
cf = cf.rename(index={0:'Actual 0',1:'Actual 1'})
cf

Predicted 0  Predicted 1
Actual 0          5.0          1.0
Actual 1          1.0         12.0

## The Opposite of accuracy is misclassification rate

In [28]:
print("The misclassification rate is:: "+str((1+1)/19))

The misclassification rate is:: 0.10526315789473684
